org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java

```diff
@@ -1277,11 +1283,11 @@ protected List<Pair<TypeBinding>> allSuperPairsWithCommonGenericType(TypeBinding
 		if (tSuper != null && s.isParameterizedType() && tSuper.isParameterizedType()) { // optimization #1 again
```

Comment:
Assuming `s.isParameterizedType` is a very cheap operation and `findSuperTypeOriginatingFrom` is not:
The lookup of tSuper can be guarded by s.isParameterizedType():
```java
TypeBinding tSuper = s.isParameterizedType() ? t.findSuperTypeOriginatingFrom(s) : null;
if (tSuper != null && tSuper.isParameterizedType()) {
  result.add(new Pair<>(s, tSuper));
}
```

Reply:
I can try this tomorrow and report back. Thank you for the hint!

Reply:
Incredible... I had to sample twice because I couldn't believe my eyes 😄 

Thank you, @szarnekow ! --> https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3411